File: python/linkedlist/leetcode/linked_list_cycle_ii.py

# 环形链表II

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    # 使用一个set
    def detectCycle_1(self, head: ListNode) -> ListNode:
        s = set()
        while head:
            if head in s:
                return head
            s.add(head)
            head = head.next

    # 快慢指针法
    def detectCycle_2(self, head: ListNode) -> ListNode:
        slow, fast = head, head
        while fast and fast.next:
            slow, fast = slow.next, fast.next.next
            if slow == fast:
                slow = head
                while slow != fast:
                    slow, fast = slow.next, fast.next
                return slow

File: go/linkedlist/leetcode/linked_list_cycle_ii.go
package main

// 环形链表II

// 使用一个set
func detectCycle1(head *ListNode) *ListNode {
	set := make(map[*ListNode]bool)
	for head != nil {
		if set[head] {
			return head
		}
		set[head], head = true, head.Next
	}
	return nil
}

func detectCycle2(head *ListNode) *ListNode {
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow, fast = slow.Next, fast.Next.Next
		if slow == fast {
			fast = head
			for slow != fast {
				slow, fast = slow.Next, fast.Next
			}
			return slow
		}
	}
	return nil
}
